lua/dap.lua

```diff
@@ -536,13 +548,19 @@ function Session:handle_body(body)
     self.message_callbacks[decoded.request_seq] = nil
     if decoded.success then
       callback(nil, decoded.body)
+      for _, c in ipairs(M.custom_response_handlers[decoded.command]) do
+        c(self, decoded.body)
+      end
```

Comment:
This is mainly to update information from session on requests that add additional information (e.g. requested scopes). So an UI can update whenever new scopes arrived. UI may also do their own requests, but this may lead to duplication of the requests.

Reply:
> A alternative would be to just have a session_information_updated event, which would UIs cause to update their information from the session instead of right after specific info requests.

Sounds also interesting, though I think that could even be done additionally - so that both lower level hooks and higher level hooks are present.

Maybe you could try out the https://github.com/mfussenegger/nvim-dap/pull/13 use case to see what works better?

Reply:
After using the new plugin for a while I observed that my that my response handlers are not always executed after the built-in ones. Maybe high-level events like session_variables_changed/session_changed might help? Or is there a way that I can be sure that nvim-dap has already handled the responses? Will `callback` only schedule the callback instead of running it immediately?

https://github.com/theHamsta/nvim-dap-virtual-text/issues/1

Reply:
Could this be caused be callbacks that in turn invoke other requests?


With a call-stack like this?

```
req1
resp1
  callback
    req2
    resp2?
    custom_callback_for_resp_2?
  custom_callback_for_resp1
```

I'm not sure if that's what is happening.

Would it be an option to move `callback(nil, decoded.body)` after the custom response handlers? To try out if this is the issue.